lib/dry/equalizer.rb

```diff
@@ -70,10 +76,17 @@ def define_cmp_method
     # @return [undefined]
     #
     # @api private
-    def define_hash_method
+    def define_hash_method(immutable:)
       keys = @keys
-      define_method(:hash) do | |
-        keys.map(&method(:send)).push(self.class).hash
+      calculate_hash = ->(obj) { keys.map { |key| obj.send(key) }.push(obj.class).hash }
+      if immutable
+        define_method(:hash) do
+          @__hash__ ||= calculate_hash.call(self)
```

Comment:
@flash-gordon is it enough? I don't see any thread related problem for this case honestly.

Reply:
it should be fine but we'll also need to override `freeze` and calculate `hash` there _before_ calling `super`, otherwise, it'll throw for frozen objects